model-server/src/test/kotlin/org/modelix/model/server/handlers/ModelReplicationServerTest.kt

```diff
@@ -0,0 +1,98 @@
+import io.ktor.http.appendPathSegments
+import io.ktor.http.takeFrom
+import io.ktor.serialization.kotlinx.json.json
+import io.ktor.server.application.install
+import io.ktor.server.plugins.contentnegotiation.ContentNegotiation
+import io.ktor.server.resources.Resources
+import io.ktor.server.routing.IgnoreTrailingSlash
+import io.ktor.server.testing.ApplicationTestBuilder
+import io.ktor.server.testing.testApplication
+import io.ktor.server.websocket.WebSockets
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.coroutineScope
+import org.modelix.authorization.installAuthentication
+import org.modelix.model.api.IConceptReference
+import org.modelix.model.client2.ModelClientV2
+import org.modelix.model.client2.readVersionDelta
+import org.modelix.model.client2.runWrite
+import org.modelix.model.client2.useVersionStreamFormat
+import org.modelix.model.lazy.RepositoryId
+import org.modelix.model.server.store.InMemoryStoreClient
+import org.modelix.model.server.store.LocalModelClient
+import kotlin.test.Test
+import kotlin.test.fail
+
+class ModelReplicationServerTest {
+
+    private fun runTest(block: suspend ApplicationTestBuilder.(scope: CoroutineScope) -> Unit) = testApplication {
+        application {
+            installAuthentication(unitTestMode = true)
+            install(ContentNegotiation) {
+                json()
+            }
+            install(WebSockets)
+            install(Resources)
+            install(IgnoreTrailingSlash)
+            val repositoriesManager = RepositoriesManager(LocalModelClient(InMemoryStoreClient()))
+            ModelReplicationServer(repositoriesManager).init(this)
+            KeyValueLikeModelServer(repositoriesManager).init(this)
+        }
+
+        coroutineScope {
+            block(this)
+        }
+    }
+
+    @Test
+    fun `pulling delta does not return objects twice`() = runTest {
+        // Arrange
+        val url = "http://localhost/v2"
+        val modelClient = ModelClientV2.builder().url(url).client(client).build().also { it.init() }
+        val repositoryId = RepositoryId("repo1")
+        val branchId = repositoryId.getBranchReference("my-branch")
+
+        // Act
+        modelClient.runWrite(branchId) { root ->
+            root.addNewChild("aChild", -1, null as IConceptReference?)
+        }
+        modelClient.runWrite(branchId) { root ->
+            root.addNewChild("aChild", -1, null as IConceptReference?)
+        }
+
+        // Assert
+        val response = client.get {
+            url {
+                takeFrom(url)
+                appendPathSegments("repositories", repositoryId.id, "branches", branchId.branchName)
+            }
+            useVersionStreamFormat()
+        }
+        val versionDelta = response.readVersionDelta()
+        val seenHashes = mutableSetOf<String>()
+        versionDelta.objectsFlow!!.collect { (hash, _) ->
+            val wasNotSeenYet = seenHashes.add(hash)
+            if (!wasNotSeenYet) {
```

Comment:
```suggestion
            val wasSeenBefore = seenHashes.add(hash)!
            if (wasSeenBefore) {
```

Maybe this version of duplicate negations is a bit easier to read?

Reply:
Done